### PR TITLE
Testing standard

### PR DESCRIPTION
**Pull request by mholmes-hs**
## Author checklist
- [x] I'm proposing a single change in this PR.
- I believe the proposed standard:
  - [x] benefits more than it costs us and
  - [x] doesn't conflict with our people-first value.
- [x] The standard's file(s) follow(s) [the established directory structure][directory-structure].
- I provided:
  - [x] a succinct summary,
  - [x] reasoning for the proposed change,
  - [x] an example/instructions on how to embody the standard unless repealing, and
  - [x] a longer explanation if appropriate.

*Check boxes if requirement met or irrelevant.*

## First reviewer checklist
***NOTE:** Some of the author's checklist items are intentionally omitted, because they are a matter of opinion and up for discussion with the whole team.*

- [x] This PR contains a single change.
- [x] The proposed standard follows [the established directory structure][directory-structure].
- The author provided:
  - [x] a succinct summary,
  - [x] reasoning for the proposed change, and
  - [x] an example/instructions on how to embody the standard unless repealing.

*Check boxes if requirement met or irrelevant.*

[directory-structure]: ../tree/master/standards#directory-structure

### CODE REVIEW

**Comment by tyleretters** [comment]
This looks great. I like the length and level of detail.

**Comment by HighlandCW** [comment]
Likes it

File: standards/practices/testing.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+2. Each test covers only one action.
+3. We formatted the test names so people can easily read it and understand what it tests.
+4. Each test is split into the standard arrange, act, and assert steps with helpful explanations.
+5. The tests cover complex functions likely affected by regressions.
+
+
+## Explanation
+Tests help us improve the quality of our software. However, tests carry maintenance costs, so **our tests should add value to our software**. In pursuit of this value:
+
+1. **Avoid overly-specific coverage goals.**
```
Comment (dlind555) [inline standards/practices/testing.md]: What about coverage goals for specific areas/libraries - e.g. 100% unit test coverage for critical custom library which handles important business logic?

File: standards/practices/testing.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+
+## Explanation
+Tests help us improve the quality of our software. However, tests carry maintenance costs, so **our tests should add value to our software**. In pursuit of this value:
+
+1. **Avoid overly-specific coverage goals.**  
+  E.g. 80% code coverage. Goals like this incentivize writing more tests than required to obtain confidence in the software.
+2. **Don't test what you don't own.**  
+  Testing languages, frameworks, and other third-party software indicates you don't trust the dependency you decided to rely on.
+3. **Avoid testing basic operations.**
```
Comment (dlind555) [inline standards/practices/testing.md]: Should we test that we're using the library correctly, e.g. for a relationship method the correct relationship id field is used and there are no typos (real-life example)? Perhaps with an integration test?

File: standards/practices/testing.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        $this->assertAuthenticatedAs($user);
+    }
+}
+```
+
+*In most frameworks, it's not worth testing authentication, because the framework handles pretty much everything. However in this example, the authentication is customized enough that we need to avoid regressions.*
+
+These two tests improve on the previous example in several ways:
+
+1. We converted from a unit test to an integration test (called a feature test in Laravel). The feature depends on several systems more simply tested with a couple short integration tests than with several unit tests that can't reliably test the whole feature.
```
Comment (dlind555) [inline standards/practices/testing.md]: Should items 1-5 be converted into set of recommendations/suggestions or should it be a case of 'learning from example'? If so, should there be more examples?

File: standards/practices/testing.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+### Unit tests
+*Test individual pieces of code.*
+
+Unit tests verify a relatively small piece of code is doing what it's intended to do. They shouldn’t have dependencies on outside systems, because they test internal consistency, not that units play nicely with outside systems.
+
+#### When (not) to use
+Write unit tests for non-trivial pieces of isolated code or to help you break integration tests into manageable steps.
+
+Avoid writing setup-heavy unit tests, because they're brittle. If you find yourself writing a unit test that requires complex mocking or requires significantly more code than the covered function, you should most likely use an integration test instead or rethink your implementation. Mocking tightly couples code to tests, which prevents proper refactoring (improving code without changing tests).
```
Comment (dlind555) [inline standards/practices/testing.md]: It might be nice to add some heuristics about how to organize the tests - e.g. when it's better to have many small tests vs. less tests with multiple assertions (and possibly setup/action/assertion blocks) in them.

File: standards/practices/testing.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+## Summary
+We test our code with as few tests as possible to provide reasonable confidence our software functions as intended.
+
+
+## Reasoning
+Testing is an essential part of software development. More specifically it:
+
+- proves our code works,
+- reduces the risk of creating regression bugs,
+- is essential for CI/CD, and
```
Comment (dlind555) [inline standards/practices/testing.md]: Should there be some information about CI/CD relationship with tests? For example, having multiple test suites - fast test suite which can be quickly run on local dev vs. the full set of tests run by CI/CD (or locally, but infrequently). Or the approaches to tests performance - using mocks or a database running in memory in a Docker container and so on.

File: standards/practices/testing.md (hunk 1 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't feel any different even for those examples. My perspective is that coverage goals under 100% are arbitrary and 100% is unrealistic and not worth the effort when accomplished.

If you have reasons why you disagree, I'd like to hear them.

File: standards/practices/testing.md (hunk 2 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't think so. If there's a bug due to a typo or something, it might be worth it, but even then, I'd generally think not.

I do think an integration test would be likely to naturally cover some of these methods w/o intentionally doing so.

File: standards/practices/testing.md (hunk 3 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I see these as recommendations through an example. I'm not entirely opposed to it explicitly saying, "These are recommendations." However, I think it's reasonably clear that when presenting reasons why something is good, those reasons are things to strive for.

I don't think there _should_ be more examples, but I do think there _could_ be more examples. I think if we want to add more examples, we should do that outside of this proposal as that would unnecessarily slow this process down.

I'd likely also want to maintain any further examples in (a) separate file(s), b/c [the standards are supposed to be short](https://github.com/HighlandSolutions/development-standards/tree/master/standards#we-prefer-fewer-simpler-standards). I hope this is the longest standard we end up with. The only reason I'm comfortable with this length is that testing is so important and complex.

File: standards/practices/testing.md (hunk 4 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't think I'm opposed to this, but I think that gets more complex and opens up more room for disagreement, so I'd prefer to see something like that added as an amendment.

The arrange, act, and assert stages are already covered in the explanation for the good test example, and I think that's sufficient.

File: standards/practices/testing.md (hunk 5 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't think so. The CI/CD relationship seems rather straightforward to me. The other stuff seems rather project-specific to me, so I'd expect the relevant team to come to an agreement when appropriate, i.e. those decisions may change over the course of a project based on things like whether test speed matters or if the project even uses a database or Docker.

File: standards/practices/testing.md (hunk 2 above)
Comment (mholmes-hs) [reply standards/practices/testing.md]: We should never test configuring our framework per its documentation.  Configuring things is supposed to be simple.  We shouldn't be trying to add maintenance costs to it.

File: standards/practices/testing.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
+# Testing
+## Summary
+We test our code with as few tests as possible to provide reasonable confidence our software functions as intended.
```
Comment (stuheiss) [inline standards/practices/testing.md]: I don't like the negative emphasis of "as few tests as possible" and suggest saying something like "We test our code with sufficient, while no more than necessary, tests to provide adequate confidence our software functions as intended"

File: standards/practices/testing.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+2. **Don't test what you don't own.**  
+  Testing languages, frameworks, and other third-party software indicates you don't trust the dependency you decided to rely on.
+3. **Avoid testing basic operations.**  
+  E.g. Testing an Eloquent model's relationships. Though you own the method on the model, in nearly every case, it merely consumes the framework.
+4. **Test the input and output, not the implementation.**  
+  Implemtation may change over time. Tests shouldn't fail after refactoring.
+
+Breaking these guidelines usually adds to maintenance costs without adding worthwhile value to our software.
+
+### Unit tests
```
Comment (stuheiss) [inline standards/practices/testing.md]: I recommend adding a paragraph about the testing pyramid of Unit, Feature, and Acceptance which is the typical way to categorize back-end tests. Then also add a bit about the pyramid of Unit, Snapshot, and e2e (end-to-end) tests that are typical of front-end and especially js apps. Edd Yerburgh in his book "Testing Vue.js Applications" talks about this on p194. Snapshot tests are a different category but  suggest noting the different pyramids too. Front-end and back-end tests are different enough to  benefit from guidance for the developer.

File: standards/practices/testing.md (hunk 6 above)
Comment (tyleretters) [reply standards/practices/testing.md]: I love the emphasis of "as few tests as possible," but understanding softening it a bit. I think the perspective is we aren't putting people in outer space or controlling robotic arms doing brain surgery.

Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't see it as negative at all. I see it as minimalist and therefore freeing. We're not saying, "don't write tests," we're saying "write as many tests as required to provide confidence, but no more."

We're also saying the same thing. I prefer the way it's written now b/c it flows better so I don't have to slow down and reread it to understand the meaning. I'm open to rephrasing it, but not at the cost of reading experience.

File: standards/practices/testing.md (hunk 7 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't find pyramids practically useful. Are we going to tally the tests in each of our categories and compare them to a pyramid? That seems valueless to me. If we're not going to do something along those lines, what purpose does including pyramids serve?

I'm opposed to adding anything we don't have sufficient experience in to a standard. I'd say given the only snapshot tests we've written are on a single project, and we're still struggling to understand them and how they can provide value, we don't have sufficient experience.

That said, we don't need a standard in order to use them, and we have a process for amending standards. I suggest we continue using them and once we have a good understanding of how to create value with them, we can amend this standard.

File: standards/practices/testing.md (hunk 6 above)
Comment (stuheiss) [reply standards/practices/testing.md]: I think we are saying the same thing but the current form leaves me with the impression that we try to avoid writing tests as much as possible rather than we write as many tests as required. I like the latter (your words) but am a bit dismayed by the former. It says to me that we don't really care about testing. I know that's not the intent but that's the message I'm getting.

File: standards/practices/testing.md (hunk 7 above)
Comment (stuheiss) [reply standards/practices/testing.md]: That sounds fine. We are inexperienced at this point. We can add more later. btw, I think the pyramid thing is helpful since it comes out of a lot of other people's experience in the larger world of testing.

Comment (curtisblackwell) [reply standards/practices/testing.md]: Yeah, I think the pyramid can be a helpful guideline when first getting started with testing, but beyond that it seems rather arbitrary to me. I don't really see a practical use for it, which is all I'm really interested in for standards.

Plus, even the pyramid is now debated within the dev community. A lot of people are pushing a "testing trophy" now. I don't really think either truly applies globally enough to form a standard around them.

File: standards/practices/testing.md (hunk 6 above)
Comment (mholmes-hs) [reply standards/practices/testing.md]: I think I prefer the way its currently worded.

File: standards/practices/testing.md (hunk 1 above)
Comment (dlind555) [reply standards/practices/testing.md]: I think that it's useful to have explicit full coverage on a specific file/library in case that this library contains critical business logic and is built to be testable from the beginning. This way it can be assured that this 'critical' functionality would perform correctly. Obviously it depends on the type of project and might only apply to a small minority of them, but I wouldn't say that such projects don't exist.

Regarding the 'arbitrariness' of test coverage goals - I'd argue that test coverage number, when it's not a goal by itself, could serve as a good indicator of how safe is it to modify the code and how likely is it that bugs in it would be caught by the tests. For example, 0-30% test coverage would indicate that the code is either doesn't require any tests, unlikely to get broken, or insufficiently tested and should be approached with caution. 30% to 80% would indicate a good amount of tests in most cases, depends on the type of code (standard boilerplate, business logic, libraries etc.), while over 80% would either indicate too many tests which would be difficult to maintain, some critical code which requires a lot of testing or accidentally high coverage without it being the intended goal.

File: standards/practices/testing.md (hunk 4 above)
Comment (dlind555) [reply standards/practices/testing.md]: What about testing methods, such as 'isPrimeNumber' (or 'shouldApplyDiscout' etc), would you have the Arrange/Act/Assert blocks as well and would have just a single call to 'isPrimeNumber' in each separate test or is it acceptable to have just one test with multiple calls to 'isPrimeNumber' and no Arrange/Act/Assert comments?

I guess my current issue is that the way it's currently structured it looks like a well-written article with a set of recommendations which is called a 'standard', without mentioning a lot of alternative approaches or going deeply into the various testing techniques. The way I see it an inexperienced developer would blindly follow it to the point as the 'one true way' after reading it (same as coding from a tutorial), while a slightly more experienced developer would receive comments about not following the standard (even when it's perfectly appropriate).

Perhaps it'd have been a better idea to heavily reduce the scope and have it much more open-ended with a list of links to learning materials at the end..

File: standards/practices/testing.md (hunk 1 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I don't disagree with those points at a high level. If there's something important enough to warrant 100% test coverage, the discussion around that would happen at the team level. This language allows for that, and [the language around overriding standards at a project level](https://github.com/HighlandSolutions/development-standards#we-put-people-first allows for that even more emphatically.

> We commit to setting aside our personal preferences and following the standards **unless our team agrees overriding a standard at the project-level benefits the people affected**.

As for your points about 30% vs 80%, that makes sense, but it's still more of an informed gut-feeling than a surety.

File: standards/practices/testing.md (hunk 4 above)
Comment (curtisblackwell) [reply standards/practices/testing.md]: I personally prefer to **always** have arrange/act/assert steps, but I'm trying to avoid prescribing my preferences too strongly.

As for not going more deeply into testing techniques, that's in an effort to adhere to the ["We prefer fewer, simpler standards" tenet](https://github.com/HighlandSolutions/development-standards/tree/master/standards#we-prefer-fewer-simpler-standards). I also believe this is much more practical. To describe techniques more deeply would require more time than I believe it's worth and make it significantly more difficult to reach a consensus. It makes more sense to me to start from a higher level where we can agree well enough on things. If it turns out things are too general to be helpful enough, we can [amend](https://github.com/HighlandSolutions/development-standards/tree/master/standards#amendment) later.

As for experienced vs. inexperienced developers, Highland currently doesn't really hire inexperienced developers, and even if we did, they'd be on a team with more experienced developers who would provide guidance. Any confusion seen there could lead to clarifying amendments.

If we just had links pointing elsewhere, that wouldn't be in our voice or control. It would also most likely require a lot of time reading things not particularly relevant to the work we do and contain conflicting info.

Basically, my goal/approach is to find the right balance of generality vs specificity to be helpful w/o losing much time to diminishing returns.

File: standards/practices/testing.md (hunk 7 above)
Comment (JHWelch) [reply standards/practices/testing.md]: In general I think these standards will help us better if we try to set them as the base level we all should be following. We shouldn't add untested, or new to us methods to a standard. It should be more of a baseline that we all understand and operate from, as opposed to as much of an inspirational goal or a possible future state. 
I think it is a sign of a problem if people are getting _introduced_ to a concept through a standard as opposed to it reinforcing ideas we are already committed to.

**Comment by mholmes-hs** [comment]
Open For Voting

**Comment by jc-web** [comment]
I like it:
- a succinct summary,
- reasoning for the proposed change, and
- an example/instructions on how to embody the standard unless repealing.

**Comment by HighlandCW** [comment]
Agrees

**Comment by JHWelch** [comment]
Voting has passed, and the procedure says I can merge it, so happy first standard everybody!
